Added default locale before login

christianrr opened the pull request:
Closes #7062

christianrr committed:
Added default locale before login

christianrr committed:
Use project_language as default language for new users

nickrum reviewed (commented):
As mentioned in the discussion, I think it would be more convenient to fall back to `project_language` when the users' language isn't set, instead of overwriting it.

christianrr commented:
> As mentioned in the discussion, I think it would be more convenient to fall back to `project_language` when the users' language isn't set, instead of overwriting it.

How can a user's language not be set? The users' language field is mandatory and whenever a new user is created in the Admin app we start off with the project_language as default - until a language is chosen by the user!? Sorry, I don't exactly understand what you mean...

christianrr committed:
Merge branch 'directus:main' into default-locale-fix7062

nickrum commented:
Sorry, I should have been more specific. As mentioned in the linked discussion, I think we should make the user's language nullable and make `null` the default value. If the user's language is set to `null`, the language of Directus should then fall back to the project language. This way the default language of all users can be changed after they have been created, unless they set their lanugage explicitly.

benhaynes commented:
I like that, good idea @nickrum! We'll want to update the interface so `NULL` shows "Project Default".

christianrr commented:
kk, I'll look into it next week!

christianrr committed:
Revert "Use project_language as default language for new users"

christianrr committed:
Allow user's language field to be nullable; Use project_language whenever language is unset;

neo7-studio-web commented:
Any chance to have it resolved in the v9 stable version ?

christianrr committed:
Apply suggestions from code review

Co-authored-by: Nicola Krumschmidt <nicola.krumschmidt@freenet.de>

christianrr committed:
Make "Project Default" option in system-language interface configurable

christianrr committed:
Set language only if user language is not null

christianrr committed:
Introduced translation for project_language field; Use translation in system-language interface

christianrr committed:
Set app language only from app.vue;
Included watcher to restore project_language if user's language is (re)set to null

christianrr committed:
Merge branch 'main' into default-locale-fix7062

christianrr commented on app/src/lang/available-languages.yaml:
```diff
@@ -1,3 +1,4 @@
+null: Project Language
```
Agreed, added all requested changes in recent commits

christianrr commented:
@nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?

benhaynes commented:
> @nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?

All PRs require multiple members of the core team to review before merging. However, the core team has been extremely busy with the release to GA (announced yesterday), the release of our new Cloud, and stability/fixes. This will get reviewed/merged, but is just not at the top of the list.

The good news is that after months of work on the tasks I mentioned, we're transitioning the team to issues and PRs full-time now... so the long delay is over and we're actively working through the list.

christianrr commented:
> > @nickrum Is there anything else todo for me? I committed the requested changes 16 days ago... Are other tasks higher prioritized at the moment, or do I miss something?
> 
> All PRs require multiple members of the core team to review before merging. However, the core team has been extremely busy with the release to GA (announced yesterday), the release of our new Cloud, and stability/fixes. This will get reviewed/merged, but is just not at the top of the list.
> 
> The good news is that after months of work on the tasks I mentioned, we're transitioning the team to issues and PRs full-time now... so the long delay is over and we're actively working through the list.

Ok, thanks for clarifying... You all do a great job, many thanks!

neo7-studio-web commented:
Any news about this one ?

rijkvanzanten commented:
> Any news about this one ?

You'd hear it here first! No activity on the PR is no updates on the PR 👍🏻

loongmxbt commented:
Same issue, where to set project default language. (login page translation)

jamescammarano committed:
Merge branch 'main' of https://github.com/directus/directus into default-locale-fix7062

jamescammarano committed:
Merge branch 'main' of https://github.com/directus/directus into default-locale-fix7062

christianrr committed:
Removed .alter() in knex migration since it's not needed for changing default and breaks checks

christianrr committed:
Merge remote-tracking branch 'origin/default-locale-fix7062' into default-locale-fix7062

christianrr committed:
Reverted previous change in migration; .alter() is needed for changing column defaults

christianrr commented:
hmm, the database migration fails to set null as default for the language field in directus_users for Oracle:
`table.string('language').nullable().alter();`

I'm not an expert on Knex migrations, how would I implement that correctly?

jaycammarano commented:
> hmm, the database migration fails to set null as default for the language field in directus_users for Oracle: `table.string('language').nullable().alter();`
> 
> I'm not an expert on Knex migrations, how would I implement that correctly?

I was going to ask @aidenfoxx

aidenfoxx commented:
@christianrr I see that `directus_users.language` is seeded as nullable. Assuming that isn't changed by any other migration then your migration is trying to set nullable to nullable, which Oracle rejects? If that's the case, I think you just need to remove that from the new migration?

aidenfoxx commented:
I don't know what the solution is then. I assumed `.defaultTo(null)` would work for Oracle. What was the error?

christianrr committed:
Reverted migration to use .defaultTo(null) to set default value

christianrr commented:
@aidenfoxx Reverted now to .defaultTo(null) - error says:
`error: alter table "directus_users" alter column "language" type varchar(255) using ("language"::varchar(255)) - ALTER COLUMN TYPE from varchar to varchar is only supported experimentally`

aidenfoxx commented:
Ah yeah... that error! Knex always generates the alter queries with the type even if you just want to change the default. And I don't think `setNullable` will work since it'll retain the existing default.

This is probably something to bring up in the https://github.com/knex/knex repo since there should be some way to remove a default!

christianrr commented:
Ok, thx for the info. I read in some forum where somebody used a raw query for altering the default value? But that wouldn't work for all database types because of different syntax, right?

Any other suggestions?

https://stackoverflow.com/questions/41959299/my-sql-alter-table-using-knex-js

aidenfoxx commented:
You could always add a Oracle specific query as we've done in the past: https://github.com/directus/directus/commit/e8e295ef6a208fcaa391ee1ce80fc67c731ebf63#diff-3bbba2b75650f2c9a7e5f1d3aa5d7bbc5b1f4e371998e122706313b93ba36edd

Or alternatively try the new `changeToTypeByCopy` helper function.

benhaynes commented:
> This is probably something to bring up in the https://github.com/knex/knex repo since there should be some way to remove a default!

Luckily we have a resident maintainor!! @rijkvanzanten 🙂

aidenfoxx commented:
@christianrr @rijkvanzanten I just had a look into this and I believe you mistook a cockroachdb error for an Oracle error. Oracle should support this correctly (though I did find and fix https://github.com/directus/directus/pull/11359 in the process 😬).

christianrr committed:
Merge branch 'directus:main' into default-locale-fix7062

christianrr committed:
Merge branch 'directus:main' into default-locale-fix7062

cr-mutor committed:
Update position of project language field in settings

cr-mutor committed:
Fix cockroachdb test error in knex migration

cr-mutor committed:
Fix cockroachdb test error in knex migration

cr-mutor committed:
Merge branch 'main' into default-locale-fix7062

# Conflicts:
#	app/src/hydrate.ts
#	app/src/interfaces/_system/system-language/system-language.vue

cr-mutor committed:
Merge remote-tracking branch 'directus/main' into default-locale-fix7062

christianrr commented:
@rijkvanzanten I tried to fix the failing migration test, however there were some recent conflicts. I have tried to merge the most recent main branch into my feature branch and resolve the conflicts, however I have the feeling that I screwed up something. My last (merge) commits are not appearing any more in the PR? Or does it now require an action from you? Sorry, I'm a bit lost ;)

cr-mutor committed:
Apply migration fix also on oracle databases

christianrr commented:
@rijkvanzanten @nickrum Everything seems to be OK now, any change to get this merged any time soon?

rijkvanzanten commented on api/src/database/migrations/20211004A-add-project-language.ts:
```diff
@@ -0,0 +1,34 @@
+		table.string('project_language').notNullable().defaultTo('en-US');
+	});
+
+	if (helper.isOneOfClients(['cockroachdb', 'oracle'])) {
+		knex.schema.raw('ALTER TABLE directus_users ALTER COLUMN language SET DEFAULT NULL');
+	} else {
```
Lets use the helpers for this, like:

https://github.com/directus/directus/blob/7d391ed4fad599240d0518fc2ab2c4bc9f20c7bf/api/src/database/migrations/20220303A-remove-default-project-color.ts#L7-L11

rijkvanzanten commented on api/src/database/migrations/20211004A-add-project-language.ts:
```diff
@@ -0,0 +1,34 @@
+	const helper = getHelpers(knex).schema;
+
+	await knex.schema.alterTable('directus_settings', (table) => {
+		table.string('project_language').notNullable().defaultTo('en-US');
```
Lets call the field `default_language` instead of `project_language` to avoid confusion with the languages configured by the individual users

rijkvanzanten commented on api/src/database/seeds/03-users.yaml:
```diff
@@ -34,7 +34,7 @@ columns:
     type: string
     length: 8
-    default: 'en-US'
+    default: null
```
We shouldn't change seeds (ever). Lets keep this on `en-US`

christianrr replied:
Ok, I'll revert this.

rijkvanzanten commented on app/src/app.vue:
```diff
@@ -85,6 +86,41 @@ export default defineComponent({
+		watch(
+			() => appStore.hydrating,
+			async () => {
+				// initially set app language after hydration
+				if (!appStore.hydrating) await setAppLanguage();
```
IIRC the app store hydrates after user login. The serverStore is hydrated _before_ the user logs in. Ideally, we set the correct language after the serverStore is hydrated, so we can show the login page in the correct language as well

christianrr replied:
Yeah, you are right. I did quite some testing however and it needs both watches: To watch the server store hydration for setting the language of the login screen and in a second stage the app hydration to set the language when the user's language is known.

rijkvanzanten replied:
> Yeah, you are right. I did quite some testing however and it needs both watches: To watch the server store hydration for setting the language of the login screen and in a second stage the app hydration to set the language when the user's language is known.

Yeah. Just curious how it ties in with the existing `setLanguage` calls, and making sure we're not doing the same thing redundantly in multiple places:

https://github.com/directus/directus/blob/ef18cdc14470aea5d6e35d0f4c1ee91aba71d132/app/src/hydrate.ts#L78

https://github.com/directus/directus/blob/ef18cdc14470aea5d6e35d0f4c1ee91aba71d132/app/src/modules/users/routes/item.vue#L438-L449

christianrr replied:
You are right, for app hydration setLanguage is called twice.

Would you prefer to have the initial setLanguage calls in hydrate.ts and server.ts (server store) or to handle it via watchers in app.vue?

rijkvanzanten commented on app/src/app.vue:
```diff
@@ -85,6 +86,41 @@ export default defineComponent({
+			}
+		);
+
+		watch(
+			() => userStore.currentUser?.language,
+			async (userLanguage) => {
+				// set project language if user's language was reset to project language (null)
+				if (!appStore.hydrating && !userLanguage) await setAppLanguage();
+			}
+		);
```
This also happens in the hydration for the usersstore if i'm not mistaken. This might be double

christianrr replied:
This is needed when the user's language is reset to Default Language on the user's detail page. However, you are right - it would be better to handle it there. I will do so.

rijkvanzanten commented on app/src/interfaces/_system/system-language/system-language.vue:
```diff
@@ -23,16 +23,24 @@ export default defineComponent({
 			type: String,
 			default: null,
 		},
+		includeProjectDefault: {
+			type: Boolean,
+			default: true,
+		},
```
Why would we want to exclude the default language from the list of options? 🤔

christianrr replied:
This is because of the settings page: `system-language` is used as interface for the Default/Project Language field. It does not make sense to be able to choose the "Default Language" option in the dropdown as value for the Default Language field.

rijkvanzanten replied:
Ohh gotcha, yeah that makes sense 👍🏻 I thought the flag would control whether or not the _language_ would show up (so like it would exclude en-US from the list f.e.). Didn't realize there was a dedicated "Project Default" option in that interface 👍🏻

rijkvanzanten reviewed (changes requested):
Thanks @christianrr! Couple quick questions:

cr-mutor committed:
Apply changes from code review

christianrr committed:
Rename migration to fix migration key error

cr-mutor committed:
Move setLanguage calls to server.ts and hydrate.ts

rijkvanzanten committed:
Merge branch 'main' into default-locale-fix7062
